dkc/core/models/quota.py

```diff
@@ -1,26 +1,62 @@
 from typing import Type
 
+from django.conf import settings
 from django.contrib.auth.models import User
-from django.db import models
+from django.core.exceptions import ValidationError
+from django.db import IntegrityError, models, transaction
 from django.db.models.signals import post_save
 from django.dispatch import receiver
-from girder_utils.models import SelectRelatedManager
+
+
+def _default_user_quota() -> int:
+    return settings.DKC_DEFAULT_QUOTA
 
 
 class Quota(models.Model):
-    allocation = models.BigIntegerField(default=0)
-    user = models.OneToOneField(User, null=True, on_delete=models.CASCADE)
-    # root_set
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(used__lte=models.F('allowed')),
+                name='used_lte_allowed',
+            )
+        ]
+
+    used = models.PositiveBigIntegerField(default=0)
+    allowed = models.PositiveBigIntegerField(default=_default_user_quota)
+
+    # OneToOneField ensures that only one Quota per User exists
+    user = models.OneToOneField(User, on_delete=models.CASCADE, related_name='quota')
+
+    @transaction.atomic
+    def increment(self, amount: int) -> None:
+        """
+        Increment or decrement the quota.
+
+        This function increments (or decrements, if ``amount`` is negative) the usage
+        values associated with this quota. If this increment would exceed the max
+        allotment for this quota, a ``ValidationError`` is raised and the transaction
+        is rolled back.
+        """
+        if amount == 0:
+            return
 
-    # The user is needed to stringify, so join it when directly querying for a quota
-    objects = SelectRelatedManager('user')
+        try:
+            # Use an .update query instead of a .save, to avoid assigning an F-expression on the
+            # local instance, which might need to be rolled back on a failure
+            Quota.objects.filter(pk=self.pk).update(used=(models.F('used') + amount))
+        except IntegrityError as e:
+            if '"used_lte_allowed"' in str(e):
```

Comment:
I wish we didn't have to do this -- do you think we'd have any luck upstreaming a feature request to psycopg2 to provide more structured info on their IntegrityErrors?

Reply:
(Not to hold up this PR or anything, just an idle thought.)

Reply:
[Here's where the `IntegrityError` is generated.](https://github.com/psycopg/psycopg2/blob/f469331af5f7b236f8037b302b69c2bf0df2a096/psycopg/error_type.c#L81-L82)

[The PostgreSQL docs state](https://www.postgresql.org/docs/current/errcodes-appendix.html):
> For some types of errors, the server reports the name of a database object (a table, table column, data type, or constraint) associated with the error; for example, the name of the unique constraint that caused a unique_violation error. Such names are supplied in separate fields of the error report message so that applications need not try to extract them from the possibly-localized human-readable text of the message. As of PostgreSQL 9.3, complete coverage for this feature exists only for errors in SQLSTATE class 23 (integrity constraint violation), but this is likely to be expanded in future.

So, it looks like it's possible in principle for only this particular case. [Psycopg 2 supports >= PostgreSQL 7.4](https://www.psycopg.org/docs/install.html#prerequisites), so support would need to be conditional.

@zachmullen I would start by filing an issue [in Psycopg 2](https://github.com/psycopg/psycopg2/issues), but if they decline it there, perhaps they'd consider it in (the apparently actively-developed) [Psycopg 3](https://github.com/psycopg/psycopg3/issues). Let me know if you do so, so I can subscribe to it.